pkg/alertmanager/state_replication_test.go

```diff
@@ -129,3 +163,212 @@ alertmanager_state_replication_total{key="nflog"} 1
+			replicationFactor: 3,
+			reads: []readStateResult{
+				{err: errors.New("Read Error 1")},
+				{err: errors.New("Read Error 2")},
+				{err: errors.New("Read Error 3")},
+			},
+			results: map[string][][]byte{
+				"key1": nil,
+				"key2": nil,
+			},
+		},
+		{
+			name:              "when reading the full state takes too long, hit timeout but become ready.",
+			replicationFactor: 3,
+			reads: []readStateResult{
+				{blocking: true},
+			},
+			results: map[string][][]byte{
+				"key1": nil,
+				"key2": nil,
+			},
+		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			reg := prometheus.NewPedanticRegistry()
+
+			replicator := newFakeReplicator()
+			replicator.reads = tt.reads
+			s := newReplicatedStates("user-1", tt.replicationFactor, replicator, log.NewNopLogger(), reg)
+
+			key1State := &fakeState{}
+			key2State := &fakeState{}
+
+			s.AddState("key1", key1State, reg)
+			s.AddState("key2", key2State, reg)
+
+			s.settleConfig.MinBackoff = 100 * time.Millisecond
+			s.settleConfig.MaxBackoff = 100 * time.Millisecond
+			s.settleConfig.MaxRetries = 3
+			s.settleReadTimeout = 1 * time.Second
+
+			assert.False(t, s.Ready())
+
+			require.NoError(t, services.StartAndAwaitRunning(context.Background(), s))
+			t.Cleanup(func() {
+				require.NoError(t, services.StopAndAwaitTerminated(context.Background(), s))
+			})
+
+			assert.True(t, s.Ready())
+
+			// Note: We don't actually test beyond Merge() here, just that all data is forwarded.
+			assert.Equal(t, tt.results["key1"], key1State.merges)
+			assert.Equal(t, tt.results["key2"], key2State.merges)
+		})
+	}
+}
+
+func TestStateReplication_GetFullState(t *testing.T) {
+
+	tc := []struct {
+		name   string
+		data   map[string][]byte
+		result *clusterpb.FullState
+	}{
+		{
+			name: "no keys",
+			data: map[string][]byte{},
+			result: &clusterpb.FullState{
+				Parts: []clusterpb.Part{},
+			},
+		},
+		{
+			name: "zero length data",
+			data: map[string][]byte{
+				"key1": {},
+			},
+			result: &clusterpb.FullState{
+				Parts: []clusterpb.Part{
+					{Key: "key1", Data: []byte{}},
+				},
+			},
+		},
+		{
+			name: "keys with data",
+			data: map[string][]byte{
+				"key1": []byte("Datum1"),
+				"key2": []byte("Datum2"),
+			},
+			result: &clusterpb.FullState{
+				Parts: []clusterpb.Part{
+					{Key: "key1", Data: []byte("Datum1")},
+					{Key: "key2", Data: []byte("Datum2")},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			reg := prometheus.NewPedanticRegistry()
+			s := newReplicatedStates("user-1", 1, nil, log.NewNopLogger(), reg)
+
+			for key, datum := range tt.data {
+				state := &fakeState{binary: datum}
+				s.AddState(key, state, reg)
+			}
+
+			result, err := s.GetFullState()
+			require.NoError(t, err)
+
+			// Key ordering is undefined for the code under test.
+			sort.Slice(result.Parts, func(i, j int) bool { return result.Parts[i].Key < result.Parts[j].Key })
```

Comment:
[nit] You don't need sorting if you use `assert.ElementsMatch()` instead of `assert.Equal()`.

Reply:
I'm checking equality the whole gRPC message, that seemed cleaner in case `FullState` ever gained new fields. I could compare just `Parts` instead? No strong opinion here really.

Reply:
Ok got it. That what you do is fine 👍